xinference/core/scheduler.py

```diff
@@ -0,0 +1,358 @@
+                        )
+                else:
+                    if r.error_msg is None:  # normal stop
+                        if not r.stream:
+                            r.future_or_queue.set_result(r.completion[0])
+                        else:
+                            await r.future_or_queue.put(XINFERENCE_STREAMING_DONE_FLAG)
+                    # Abnormal stop, currently indicates that the parameter check does not pass,
+                    # and does not participate in the inference
+                    else:
+                        stop_before_prefill_cnt += 1
+                        if not r.stream:
+                            r.future_or_queue.set_exception(ValueError(r.error_msg))
+                        else:
+                            await r.future_or_queue.put(
+                                XINFERENCE_STREAMING_ERROR_FLAG + r.error_msg
+                            )
+
+        force_clear_cache: bool = False
+        if len(self._running_queue) > 0:
+            batch_size_after_one_step = len(self._running_queue)
+            batch_size_before_one_step = batch_size - stop_before_prefill_cnt
+            if batch_size_before_one_step != batch_size_after_one_step:
+                assert batch_size_after_one_step < batch_size_before_one_step
+                force_clear_cache = True
+                for r in self._running_queue:
+                    r.kv_cache = None
+
+        # clean cuda cache
+        if (
+            force_clear_cache
+            or self._iter_cnt % XINFERENCE_BATCHING_CLEAN_CACHE_INTERVAL == 0
+        ):
+            """
+            Clear unused kv_cache.
+            - force_clear_cache=True indicates that some requests quit and their kv_cache can be cleared.
+            - self._iter_cnt means that no need to empty the cache every step.
+            """
+            gc.collect()
+            empty_cache()
+
+    async def add_request(self, prompt: str, future_or_queue, *args, **kwargs):
+        req = InferenceRequest(prompt, future_or_queue, True, *args, **kwargs)
+        rid = req.request_id
+        if rid is not None:
+            if rid in self._id_to_req:
+                raise KeyError(f"Request id: {rid} has already existed!")
+            self._id_to_req[rid] = req
+        self._waiting_queue.append(req)
+
+    async def abort_request(self, req_id: str) -> str:
+        """
+        Abort a request.
+        Abort a submitted request. If the request is finished or not found, this method will be a no-op.
+        """
+        if req_id not in self._id_to_req:
+            logger.info(f"Request id: {req_id} not found. No-op for xinference.")
+            return AbortRequestMessage.NOT_FOUND.name
+        else:
+            self._abort_req_ids.add(req_id)
+            logger.info(f"Request id: {req_id} found to be aborted.")
+            return AbortRequestMessage.DONE.name
+
+    async def run(self):
+        while True:
+            await self.step()
```

Comment:
We'd better add a try-catch block and logs in the while loop to avoid the thread exiting silently. For example, if an exception is raised, it will cause the thread to exit.